### DISCUSSION TITLE
customising-the-new-telemetry-logging-source-generator/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#399] [Announcements]
# customising-the-new-telemetry-logging-source-generator/

In this post I look at the [TagProperties] and ILogEnricher features from the Microsoft.Extensions.Telemetry and show how they can be used in your apps.

https://andrewlock.net/customising-the-new-telemetry-logging-source-generator/

**Comment by CheloXL** [comment]
Is there a way to enrich a log with data from httpcontext? By what I saw, all enrichers are registered as singleton. I would like to have a way to add some custom tracing information to all my logs that start from an http request...

**Reply by andrewlock** [reply]
Yeah, that shouldn't be a problem - you should be able to implement `ILogEnricher`, and use the `IHttpContextAccessor` to access the HttpContext I believe

**Reply by CheloXL** [reply]
The problem I see with this is that enrichers are registered as singleton, so `IHttpContextAccessor` would not be available since there is no context...

**Reply by tkrafael** [reply]
As long as logging is called inside http request, the context is available inside log enricher. However you must guard against null.
Also, depending when the enricher is run, request and response might not be completed. Example: you want to add status code to log when logging is called before response is processed.
My suggestion is to enrich using endpoint filters or Middleware. That makes code clearer on its purpose.

**Comment by tkrafael** [comment]
When using net8 it is possible to log properties from complex objects using the following string:
`"user contains {@order}"` where the object passed in the logger message must have same name.

**Reply by CheloXL** [reply]
Yes, I know that. But what I want it to enrich my entries with the value from some custom header.

**Comment by voroninp** [comment]
Typo, the second should be just ILogEnricher

`IStaticLogEnricher` — enrichers are created once on app startup, and their tags added to the logger. Use if the values you're adding won't change for the app lifetime.

`IStaticLogEnricher` —enrichers are invoked every time you write a log, so good for values that may

**Comment by voroninp** [comment]
I guess enrichment won't work for Serilog, because it replaces logger factory, so there won't be any `ExtendedLoggerFactory`?

**Reply by tkrafael** [reply]
You can receive IHttpContextAccessor as singleton, no worries.
However you have to check if Context is not null. It will be null whenever a log event occurs outside an http context

**Comment by mmartijnNL** [comment]
I added a StaticLogEnricher. It works fine for my own log lines. But the logs from MS for Azure Functions do not get enriched. I'm talking about the default logs that go "Executing 'FunctionName' (Reason='This function was programmatically called via the host APIs.', Id=blablabla)". I still see these logs in application insights, but without the enriched property. Is there any way to get these properties in those logs?
